### PR TITLE
feat: add support Application Default Credentials for google-cloud-storage

### PR DESCRIPTION
**Pull request by nakatanakatana**
Tihs PR adds setting to use Application Default Credentials for authentication of google-cloud-storage.

reference:

* https://cloud.google.com/docs/authentication/application-default-credentials
* https://googleapis.dev/nodejs/storage/latest/Storage.html#Storage-examples

### CODE REVIEW

File: src/env.ts [hunk 1]
```diff
@@ -40,6 +40,7 @@ const schema = Type.Object(
     GCS_PROJECT_ID: Type.Optional(Type.String()),
     GCS_CLIENT_EMAIL: Type.Optional(Type.String()),
     GCS_PRIVATE_KEY: Type.Optional(Type.String()),
+    GCS_USE_ADC: Type.Optional(Type.String({ default: 'false' })),
```
Comment (fox1t) [inline src/env.ts:43]:
```suggestion
    GCS_USE_ADC: Type.Optional(Type.Boolean({ default: false })),
```

File: src/plugins/remote-cache/index.ts [hunk 2]
```diff
@@ -58,6 +58,7 @@ async function turboRemoteCache(
       clientEmail: instance.config.GCS_CLIENT_EMAIL,
       privateKey: instance.config.GCS_PRIVATE_KEY,
       projectId: instance.config.GCS_PROJECT_ID,
+      useApplicationDefaultCredentials: instance.config.GCS_USE_ADC == 'true',
```
Comment (fox1t) [inline src/plugins/remote-cache/index.ts:61]:
```suggestion
      useApplicationDefaultCredentials: instance.config.GCS_USE_ADC === true,
```

**Review by fox1t — APPROVED** [review]
Would you mind adding tests for this configuration?

**Review by fox1t — CHANGES_REQUESTED** [review]
Please look at suggestions

**Comment by emalihin** [comment]
Alternatively we could use ADC if no clientEmail || privateKey` are provided?

**Comment by fox1t** [comment]
> Alternatively we could use ADC if no `clientEmail || privateKey` are provided?

Would you mind elaborating? I am not an expert when it comes to GCP.

**Comment by emalihin** [comment]
> Would you mind elaborating? I am not an expert when it comes to GCP.

From docs:
> ADC searches for credentials in the following locations:
>    [GOOGLE_APPLICATION_CREDENTIALS environment variable](https://cloud.google.com/docs/authentication/application-default-credentials#GAC)
>    [User credentials set up with the Google Cloud CLI](https://cloud.google.com/docs/authentication/application-default-credentials#personal)
>    [The attached service account, as provided by the metadata server](https://cloud.google.com/docs/authentication/application-default-credentials#attached-sa)

By simply not providing credentials to `Storage()` constructor we instruct it to use one of the above three methods. The 2nd and 3rd scenarios are most common in my experience - SDK knows where to find local credentials, and it will also check the GCP metadata server for the Service Account information (this is used for Workload Identity auth in GKE - the recommended way to authenticate with GCP. Very similar to AWS's `IAM roles for service accounts/IRSA`)

**Comment by tehKapa** [comment]
@all-contributors add @nakatanakatana for code

**Comment by fox1t** [comment]
This PR was superseded by https://github.com/ducktors/turborepo-remote-cache/pull/68

**Pull request closed without merging**